File: JS/Tree/894. All Possible Full Binary Trees.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number} n
 * @return {TreeNode[]}
 */
var allPossibleFBT = function(n) {
    if (n % 2 === 0) {
        return [];
    } else if (n === 1) {
        return [new TreeNode(0)];
    } else {
        const result = [];
        for (let i = 1; i < n; i+=2) {
             allPossibleFBT(i).forEach(j => { // left tree 1 ~ n - 2
                  allPossibleFBT(n - i - 1).forEach(k => { //right tree 1 ~ n - 2
                      const root = new TreeNode(0);
                      root.left = j;
                      root.right = k;
                      result.push(root);
                  });
             });
        }
        return result;
    }
};


var allPossibleFBT = function(n) {  // DP solution
    if (n % 2 === 0) {
        return [];
    }
    const dp = [[], [new TreeNode(0)]];
    for (let i = 3; i <= n; i += 2) {
        dp[i] = [];
        for (let j = 1; j < i; j += 2) {
            dp[j].forEach(l => {
                dp[i - j - 1].forEach(m => {
                    const node = new TreeNode(0, l, m);
                    dp[i].push(node);
                });
            });
        }
    }
    return dp[n];
};

File: Golang/Tree/894.all-possible-full-binary-trees.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func allPossibleFBT(n int) []*TreeNode {
	cache := map[int][]*TreeNode{}
	var helper func(n int) []*TreeNode
	helper = func(n int) []*TreeNode {
		ret := []*TreeNode{}
		if n == 1 {
			ret = append(ret, &TreeNode{})
		} else if n%2 == 1 {
			if value, ok := cache[n]; ok {
				return value
			}
			for i := 1; i < n; i += 2 {
				left := helper(i)
				right := helper(n - 1 - i)
				for _, j := range left {
					for _, k := range right {
						root := &TreeNode{
							Left:  j,
							Right: k,
							Val:   0,
						}
						ret = append(ret, root)
					}
				}
			}
			cache[n] = ret
		}
		return ret
	}
	return helper(n)
}

func allPossibleFBTDP(n int) []*TreeNode {
	dp := make(map[int][]*TreeNode, n+1)
	dp[1] = []*TreeNode{&TreeNode{}}
	for i := 2; i <= n; i++ {
		if i%2 == 0 {
			dp[i] = []*TreeNode{}
		} else {
			nodes := []*TreeNode{}
			for j := 1; j < i; j += 2 {
				left := dp[j]
				right := dp[i-1-j]
				for _, k := range left {
					for _, l := range right {
						nodes = append(nodes, &TreeNode{
							Left:  k,
							Right: l,
						})
					}
				}
			}
			dp[i] = nodes
		}
	}
	return dp[n]
}
